### ISSUE TITLE
High CPU load while performing disk R/W

### ISSUE THREAD

**Issue by protegio** [veracrypt/VeraCrypt#259]
Hi,

I've noticed that while reading or writing in an encrypted volume (AES-SHA256) the CPU usage spikes up to 40%. It was tested on multiple combinations of the following configurations:

* VeraCrypt Version: 
  * 1.17 
  * 1.21
* OS:
  * Windows XP 
  * Windows 7
* Machine:
  * 45nm Core 2 Duo with 2gb of RAM (40 % CPU usage)
  * 3rd gen i5 with 4gb of ram (25% CPU usage)
  * Virtual Machine

I have also benchmarked these operations with no encryption at all and with DiskCryptor encryption.  In both cases the CPU usage never went over 3%.

Is this a known issue? Is there a patch or configuration to improve the CPU usage?

Thank you.

**Comment by protegio** [comment]
Running the tests again I noticed that Veracrypt uses 100% of one of the CPU's cores when reading from an encrypted system partition. If the CPU has two cores I get near 50% CPU usage, 25% for 4 cores and 13% for 8 cores.

I'm already looking in the Veracrypt's code, trying  to find which function is consuming more CPU. I used the Xperf to profile the driver and I got high CPU usage in the driver's crypto thread. Apparently, in the spin lock manipuation functions. But I couldn't find exactly where.

As this is my first time using  Xperf, probably there is room for improvement in the Xperf configuration. I also don't know Veracrypt's code very well, so I may be looking in the wrong place.

Just to let you know, I repeated my tests using Bitlocker and the CPU usage never went over 5%. So I really believe this is an issue with Veracrypt. Unless the CPU usage comes from a security feature not implemented by DiskCryptor and Bitlocker.

**Comment by protegio** [comment]
I found this piece of code in the `EncryptionThreadProc` function that doesn't look right. Apparently, the while loop may iterate multiple times before its condition is met, like a busy wait. Why not use a separated "ready" event for each work item? 

You can find multiple occurrences of this logic in the `EncryptionThreadPool.c` file.

```cpp
TC_ACQUIRE_MUTEX (&DequeueMutex);

workItem = &WorkItemQueue[DequeuePosition++];

if (DequeuePosition >= TC_ENC_THREAD_POOL_QUEUE_SIZE)
	DequeuePosition = 0;

while (!StopPending && GetWorkItemState (workItem) != WorkItemReady)
{
	TC_WAIT_EVENT (WorkItemReadyEvent);
}

SetWorkItemState (workItem, WorkItemBusy);

TC_RELEASE_MUTEX (&DequeueMutex);
```

**Comment by protegio** [comment]
Running Xperf again I found that the function [`KeSaveFloatingPointState`](https://msdn.microsoft.com/en-us/library/windows/hardware/ff553243(v=vs.85).aspx), called by [`DecipherBlocks`](https://github.com/veracrypt/VeraCrypt/blob/ea94096df71a7100d3c0f707898d9760f624da32/src/Common/Crypto.c#L325) and [`EncipherBlocks`](https://github.com/veracrypt/VeraCrypt/blob/ea94096df71a7100d3c0f707898d9760f624da32/src/Common/Crypto.c#L208)  (`Common/Crypto.c`), is responsible for the high CPU usage in machines with crypto hardware acceleration.

If I remove this function call from the code, the CPU usage is reduced from 25% to 5~10%.  However, from my understanding, this function must be called before calling the CPU's crypto instructions, to avoid messing with other threads using the floating point registers.

As DiskCryptor also calls `KeSaveFloatingPointState` before performing the encryption and doesn't use as much CPU as VeraCrypt, I think there are two possibilities:

* VeraCrypt calls `KeSaveFloatingPointState` more times than DiskCryptor.
* and/or VeraCrypt is running in a lower priority level that hangs the KeSaveFloatingPointState calls in a spin lock.

This week I'll try to mess with these variables and see how VeraCrypt behave.  I let you know if I have any progress (if there is someone reading this).

**Comment by Lyoko-Jeremie** [comment]
i think it seems like have some releationship to #136  and  #208 .

thank you for your work @protegio , and have any discovery now ?

**Comment by protegio** [comment]
> i think it seems like have some releationship to #136 and #208 .

I don't think so, because the KeSaveFloatingPointState are called only in 32 bits architecture. I believe these issues were found in 64 bits computers.

> thank you for your work @protegio , and have any discovery now ?

I was able to reduce the CPU usage to 15% (10% reduction) by moving the KeSaveFloatingPointState to a higher scope (reducing the number of calls to this function). This number is closer to what I expected in comparison to DiskCryptor, which has 2x faster crypto functions and uses only 5% of the CPU while transferring big files.

As DiskCryptor has its XTS implemented in assembly - while VeracCrypt implements it in C/C++ -, and has a simpler IRP management, I don't think I can improve the CPU usage furthermore, without major changes in the code.

Also, I found that HDTune's default benchmark is not reliable to compare DiskCryptor to VeraCrypt. I recommend using the file benchmark.
